### ISSUE TITLE
Make Link error from Client to Gateway (invalid docker container)

### ISSUE THREAD

**Issue by g0lg0-93** [fractalnetworksco/selfhosted-gateway#49]
hey team, first off thanks for putting this proj together. It reads and looks promising! 

I've configured my VPS as instructed, no errors received there. However, I have reached the 'make link' part of building the client server and I have entered the 'expose' portion exactly as it states (EXPOSE=nextcloud:443) in the instructions, but my containers apparently invalid. I'm trying to expose my nextcloud container, which is on 443. 

I've also opt'd to set my the VPS to have a wildcard record - if that makes any difference.

docker-compose below for nextcloud
`---
services:
  nextcloud:
    container_name: nextcloud
    image: lscr.io/linuxserver/nextcloud:latest
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=Etc/UTC
    volumes:
      - /home/whoami/nextcloud:/config
      - /Volume/mnt:/data
    ports:
      - 443:443
    restart: unless-stopped
  collabora:
     image: collabora/code:latest
     container_name: collabora
     cap_add:
      - MKNOD
`

Let me know if you need any more snippets or further clarification on my setup.

**Comment by zacharylott94** [comment]
Is `make link` generating a docker compose snippet for you or are you getting some error at that step? If it *is* generating the snippet, is the problem that you just can't reach nextcloud from your VPS? I've spun up a nextcloud container to follow along with, and uncommenting the caddy options in the generated link snippet seems to allow me to get to my nextcloud through a VPS. Commenting out `TLS_INTERNAL` also gave me a secure connection when before, I was getting browser warnings.
![image](https://github.com/fractalnetworksco/selfhosted-gateway/assets/37595873/f1efe274-7512-4da0-921a-feacfc2b2128)
I couldn't reach nextcloud before uncommenting those lines.
Hope that helps. Reach out again if I missed the mark.

**Comment by g0lg0-93** [comment]
_It would've helped if I had posted the error I received :)_ 
Below is what I receive on the client. I receive a snippet if I remove the port I want to expose, but I then have to fill in the blanks on the snippet (exposed will be blank & gateway_endpoint will be blank)
```
make link GATEWAY=user@x.x.x.x FDQN=mydomain.com EXPOSE=nextcloud:443
docker run -e SSH_AGENT_PID=$SSH_AGENT_PID -e SSH_AUTH_SOCK=$SSH_AUTH_SOCK -v $SSH_AUTH_SOCK:$SSH_AUTH_SOCK -v "$PWD:/workdir" --rm -it fractalnetworks/gateway-cli:latest user@x.x.x.x  nextcloud:443
user@0.0.0.0's password: 
docker: Error response from daemon: Invalid container name (nextcloud:443), only [a-zA-Z0-9][a-zA-Z0-9_.-] are allowed.
See 'docker run --help'.
make: *** [Makefile:16: link] Error 125
```

**Comment by zacharylott94** [comment]
Hmm, this may be unrelated, but having to enter my ssh password never worked out for me for some reason. To bypass that, I add my ssh key while my working directory is the selfhosted-gateway folder with `ssh-add /.ssh/id_rsa` and then use `make link`. It no longer prompts for password when I do it that way. 
Should this change anything about how the scripts run? No, but I've found that computers can be temperamental. :)
If you want to try that out, I'll continue looking into the problem in the meantime.

**Comment by zacharylott94** [comment]
Uh, try to add a subdomain here as well: `make link GATEWAY=user@x.x.x.x FDQN=mydomain.com EXPOSE=nextcloud:443`
For example: `make link GATEWAY=user@x.x.x.x FDQN=nextcloud.mydomain.com EXPOSE=nextcloud:443`

**Comment by zacharylott94** [comment]
Oh, and also, make sure you delete any old link containers on your VPS as you're testing things. `make link` doesn't trash them or rebuild them or anything. So, if you successfully generate a snippet, there's already a corresponding link container on the VPS side.

**Comment by zacharylott94** [comment]
So, if you don't have a subdomain in your FQDN, the script doesn't grab the resolving IP, which causes the resulting snippet to not put the IP of your VPS in `GATEWAY_ENDPOINT`. It ends up looking like what you have in that field where it's just the port. `GATEWAY_ENDPOINT: :32769`. I'm not sure if that's the entirety of the problem though.

**Comment by g0lg0-93** [comment]
I'd like to thank you for your help.

I got frustrated and walked away. The VPS server had nothing special on it, so I nuked it and started from scratch. It worked first time. One thing I done differently was give the same username I have on the client host when creating a 2nd user on the gateway server, which i hadn't done last time. Also, I had installed wireguard on the server prior to going down the 'selfhosted-gateway' route. I can reach the server publicly, so I can now move on with my life.
